File: src/main/java/june_2025/PermutationInString.java
package june_2025;

public class PermutationInString {

    public static void main(String[] args) {
//        String s1 = "abcg";
//        String s2 = "eidbaocaboo";

        String s1 = "adc";
        String s2 = "dcda";

        PermutationInString permutationInString = new PermutationInString();
        boolean result = permutationInString.checkInclusion(s1, s2);
        System.out.println(result);

    }

    // O(n) time | O(26) space
    public boolean checkInclusion(String s1, String s2) {
        int[] pattern = new int[26];
        for (char c : s1.toCharArray()) {
            pattern[c - 'a']++;
        }
        int[] substr = new int[26];

        for (int i = 0; i < s2.length(); i++) {
            char c = s2.charAt(i);
            if (i >= s1.length()) {
                boolean found = true;
                for (int j = 0; j < 26; j++) {
                    if (pattern[j] != substr[j]) {
                        found = false;
                        break;
                    }
                }
                if (found) {
                    return true;
                }
                char t = s2.charAt(i - s1.length());
                substr[t - 'a']--;
            }
            substr[c - 'a']++;
        }
        for (int i = 0; i < 26; i++) {
            if (pattern[i] != substr[i]) {
                return false;
            }
        }
        return true;
    }


}

File: python/permutation_in_string.py
class Solution(object):
    # O(n) time | O(26) space
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        pattern = [0] * 26
        for i, c in enumerate(s1):
            pattern[ord(c) - ord('a')] += 1
        
        substr = [0] * 26

        for i in range(len(s2)):
            c = s2[i]
            if i >= len(s1):
                found = True
                for j in range(26):
                    if pattern[j] != substr[j]:
                        found = False
                        break
                
                if found:
                    return True
                
                t = s2[i - len(s1)]
                substr[ord(t) - ord('a')] -= 1

            substr[ord(c) - ord('a')] += 1
        
        for j in range(26):
            if pattern[j] != substr[j]:
                return False
        return True
        



solution = Solution()

s1 = "abc"
s2 = "eidbaocaboo"

print(solution.checkInclusion(s1, s2))